lib/deimos/active_record_consume/batch_consumption.rb

```diff
@@ -116,6 +116,23 @@ def upsert_records(messages)
         import_associations(valid_upserts) unless @association_list.blank?
       end
 
+      def remove_associations(assoc, keys)
+        foreign_keys = keys.map { |key| key[assoc.foreign_key] }.compact
+        associated_keys = keys.map { |key| key[associated_column] }.compact
+        if foreign_keys.length.zero? && associated_keys.length.zero?
+          return
+        end
+
+        assoc.klass.unscoped.where.not(
+          assoc.foreign_key => foreign_keys,
+          associated_column => associated_keys
+        ).delete_all
```

Comment:
Oof this isn't going to work. Imagine if we have the following:

FK1->AC1  
FK2->AC1  
FK3->AC3  
FK4->AC4

and we have the following messages coming in:  
  
FK1->AC1

What this query is saying is:

"if a record doesn't have any of the given foreign keys (FK1) and doesn't have any of the given association keys (AC1) delete it"

This would delete every row in the database besides the first one! :grin: 

The only way we could do this would be to do a query one FK (i.e. one offer\_id) at a time, or to join it together in a giant WHERE clause... neither of which are great :thinking: 

I apologize for the fuzzy thinking. We'll need to think this through a bit more carefully.